torch/testing/_internal/common_utils.py

```diff
@@ -2439,6 +2440,265 @@ def genSparseTensor(self, size, sparse_dim, nnz, is_uncoalesced, device, dtype):
+        """Generator of simple inputs for tensor constructors of the given layout.
+
+        The generated tensor inputs have the following properties:
+
+        - tensor shapes are minimal but not trivial
+        - tensor values are sorted sequences for COO and CSR formats, e.g. [1, 2, 3, 4]
+        - the generated tensors represent the same mathematical tensor for all layouts
+        - the generated tensors include regular, zero-sized, and optionally, batched or/and hybrid tensors.
+
+        If output_tensor is True, yield tensors with the given
+        layout. Otherwise, yield inputs to the corresponding tensor
+        constructors:
+
+          - sparse compressed input is defined as
+            (compressed_indices, plain_indices, values), dict(size=expected_size_from_shape_inference, device=device, dtype=dtype)
+
+          - sparse COO input is defined as
+            (indices, values), dict(size=expected_size_from_shape_inference, device=device, dtype=dtype)
+
+          - strided input is defined as
+            (values,), dict(device=device, dtype=dtype)
+        """
+        if index_dtype is None:
+            index_dtype = torch.int64
+
+        is_compressed_sparse_layout = layout in {torch.sparse_csr, torch.sparse_csc, torch.sparse_bsr, torch.sparse_bsc}
+
+        if output_tensor:
+            for args, kwargs in self.generate_simple_inputs(layout, device=device, dtype=dtype, index_dtype=index_dtype,
+                                                            enable_batched=enable_batched, enable_hybrid=enable_hybrid,
+                                                            enable_zero_sized=enable_zero_sized,
+                                                            enable_batched_variable_nse=enable_batched_variable_nse,
+                                                            output_tensor=False):
+                if layout is torch.strided:
+                    size = kwargs.pop('size', args[0].shape)  # to ensure that a zero-sized tensor has the desired shape
+                    assert len(args) == 1
+                    yield args[0].reshape(size)
+                elif layout is torch.sparse_coo:
+                    yield torch.sparse_coo_tensor(*args, **kwargs).coalesce()
+                elif is_compressed_sparse_layout:
+                    kwargs.update(layout=layout)
+                    yield torch.sparse_compressed_tensor(*args, **kwargs)
+                else:
+                    assert 0  # unreachable
+            return
+
+        def expand_values(base, densesize):
+            """
+            Expand values in base to blocks of values with densesize.
+            """
+            if not densesize:
+                return base
+            if isinstance(base, list):
+                return [expand_values(b, densesize) for b in base]
+
+            r = 0
+            for s in reversed(densesize):
+                r = [copy.deepcopy(r) for _ in range(s)]
```

Comment:
Hmm, why can't we just create `r` as `torch.zeros(densesize)`?

Reply:
It does not work with tensors directly? Also, the Tensor class has a `tolist` method that preserves it's nesteness.

Reply:
I am not sure about that, now after looking at the test... And I am not sure we ever construct sparse tensors from something which is not TensorLike in ATen...

Reply:
You can construct tensors from lists, for instance:
```python
>>> torch.sparse_csr_tensor([0, 2, 4], [0, 1, 0, 1], [1, 2, 3, 4])
tensor(crow_indices=tensor([0, 2, 4]),
       col_indices=tensor([0, 1, 0, 1]),
       values=tensor([1, 2, 3, 4]), size=(2, 2), nnz=4,
       layout=torch.sparse_csr)
```
In tests and examples, we often construct tensors from Python lists, so I don't see the relevance of constructing these within ATen.

Reply:
If using tensors leads to cleaner code, and when there is `tolist`, then why not?

Reply:
> What about we create r as torch.zeros(densesize)? It appears that the output of this function is fed into torch.tensor anyway...

Done.